tron/kubernetes.py

```diff
@@ -135,8 +135,42 @@ def handle_event(self, event: Event) -> None:
         if k8s_type == "running":
             self.started()
         elif k8s_type == "finished":
-            self.exited(0)
+            raw_object = getattr(event, "raw", {})
+            container_statuses = raw_object.get("status", {}).get("container_statuses", [])
+            abnormal_exit = False
+
+            if len(container_statuses) > 1:
+                # shouldn't happen right now, but who knows what future us will do :p
+                self.log.error(
+                    "Got an event for a Pod with multiple containers - not inspecting payload to verify success."
+                )
+            else:
+                main_container_statuses = container_statuses[0]
+                main_container_state = main_container_statuses.get("state")
+                if main_container_state is None or main_container_state.get("terminated") in None:
+                    self.log.error("Got an event with missing state - assuming success.")
+                else:
+                    termination_metadata = main_container_state.get("terminated")
+                    # this is kinda wild: we're seeing that a kubelet will sometimes fail to start a container (usually
+                    # due to what appear to be race conditons like those mentioned in
+                    # https://github.com/kubernetes/kubernetes/issues/100047#issuecomment-797624208) and then decide that
+                    # these Pods should be phase=Succeeded with an exit code of 0 - even though the container never actually
+                    # started. So far, we've noticed that when this happens, the finished_at and reason fields will be None
+                    # and thus we'll check for at least one of these conditions to detect an abnormal exit and actually "fail"
+                    # the affected action
+                    # NOTE: hopefully this won't
+                    abnormal_exit = (
+                        termination_metadata.get("exit_code") == 0 and termination_metadata.get("finished_at") is None
```

Comment:
maybe we can also check for reason as well? Something like
`termination_metadata.get("exit_code") == 0 and (termination_metadata.get("finished_at") is None or termination_metadata.get("reason") is None)`

Reply:
I think we'd want `and` instead of `or` if we wanna match what we've been seeing exactly